### ISSUE TITLE
Conventional commits do not respect '!' on types

### ISSUE THREAD

**Issue by alexforsyth** [lerna/lerna#2668]
Lerna conventional commits do not respect '!' as denoting a breaking change when part of the 'types' on a [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#examples). Additionally, It seems like including the '!' may make lerna not recognize that commit as a valid conventional commit in some cases. [This repo](https://github.com/alexforsyth/lerna-conventional-commits) also highlights this problem. 

Repro repo: 
https://github.com/alexforsyth/lerna-conventional-commits

## Expected Behavior

The generated changelog should list breaking changes, when a '!' is included on a type. 

ex. 2 from [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#examples)
```
refactor!: drop support for Node 6
```
Should be equal to using the `BREAKING CHANGE: ` phrase in the footer.

## Current Behavior

The generated changelog does not include the change when '!' is included on a type. 

ex.
```
refactor!: drop support for Node 6
```

## Steps to Reproduce (for bugs)
[This repo is a live example](https://github.com/alexforsyth/lerna-conventional-commits) that highlights this problem.

**Comment by alexforsyth** [comment]
Bump

**Comment by kamsar** [comment]
The default settings uses [conventional-changelog-angular](https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-angular) which is not the conventional commits spec, and doesn't support !.

It will work if you add `--changelog-preset conventionalcommits`, see https://github.com/lerna/lerna/issues/2138

Also in your repro, you're explicitly specifying the bump as `minor`, which overrides using conventional commits to determine the desired bump - maybe that's what you want, though.

https://github.com/alexforsyth/lerna-conventional-commits/blob/b575a9278be8ea053a8e1617014fe9be679d1272/package.json#L12

**Comment by kamsar** [comment]
This can also be specified in `lerna.json`: https://github.com/lerna/lerna/tree/master/commands/version#--create-release-type

**Comment by mrchief** [comment]
I'm seeing the same issue with lerna 3.2 and v4. Adding 

```
{
  "changelogPreset": {
    "name": "conventionalcommits",
    "issueUrlFormat": "{{host}}/{{owner}}/{{repository}}/issues/{{id}}"
  }
}
```

has no effect.

**Comment by mrchief** [comment]
Digging further I found this: 
![image](https://user-images.githubusercontent.com/781818/110262742-9c519380-7f82-11eb-9240-15b6726f7ca1.png)

which is the right behavior as `button` has breaking changes. 

which seems to be related to https://github.com/lerna/lerna/issues/2761, https://github.com/lerna/lerna/issues/2766

After upgrading packages to `1.x`, versioning seems to do the right thing (respect `feat!:...`)

![image](https://user-images.githubusercontent.com/781818/110262963-44fff300-7f83-11eb-8dcb-07294b8d42cb.png)

**Comment by xiaoxinghu** [comment]
how do we get major version bump to work? I am using v4 and I tried with commit messages have both `!` and/or `BREAKING CHANGE:` all over the place, I still get minor. Here is my Lerna config:
```
{
  "version": "2.6.0",
  "npmClient": "yarn",
  "useWorkspaces": true,
  "command": {
    "version": {
      "allowBranch": "master",
      "conventionalCommits": true,
      "createRelease": "github",
      "exact": true,
      "message": "chore(release): %s"
    }
  },
  "ignoreChanges": ["**/__tests__/**", "**/*.test.ts"],
  "packages": ["packages/*"]
}

```
and I use `lerna publish --yes` in ci.

**Comment by privatenumber** [comment]
If your package is in major version zero, the semver rules are shifted over, so _breaking changes_ happen in 0.x.0, and _features & patches_ happen in 0.0.y.

Therefore, a breaking change in a package that is in 0.x.y, will only bump _x_.

I wish this was better documented, but it's in the [npm docs](https://docs.npmjs.com/cli/v6/using-npm/semver#:~:text=%5E0.2.3%20%3A%3D%20%3E%3D0.2.3%20%3C0.3.0):
> ^1.2.3 := >=1.2.3 <2.0.0
> ^0.2.3 := >=0.2.3 <0.3.0
> ^0.0.3 := >=0.0.3 <0.0.4

**Comment by richardkazuomiller** [comment]
I also can't get `BREAKING CHANGE` to work in my project. Has anyone figured this out? 🤔

**Comment by privatenumber** [comment]
@richardkazuomiller

Provide your lerna configuration and the `package.json` of your subpackage so others can help

**Comment by richardkazuomiller** [comment]
lerna.json

```
{
  "packages": [
    "packages/*"
  ],
  "version": "1.23.0",
  "command": {
    "publish": {
      "conventionalCommits": true,
      "message": "chore(release): publish",
      "registry": "https://npm.pkg.github.com",
      "allowBranch": "master"
    }
  }
}
```

I don't think there's anything of relevance in package.json

**Comment by privatenumber** [comment]
Maybe. Sometimes, the bug is in the places you least suspect (which is why you don't find it 😅).

In this case, I was curious what the version of your subpackage is. But seems youre not using independent mode.

**Comment by richardkazuomiller** [comment]
I tried a few things and finally got the major version bump 😂.

Doing an empty commit (`git commit --allow-empty ...`) like this looks like it should work when not using independent mode, but doesn't:

```
feat(my-package): aaa

BREAKING CHANGE: bbb
```

Putting `BREAKING CHANGE` in a merge commit, somehow is recognized by conventional commits, shown by the fact that it shows up in the changelog, however it also doesn't bump the version.

What finally worked is we reworded a recent commit to a file inside one of the packages to include `BREAKING CHANGE` in the description of the commit. For people like me who prefer to think in terms of PRs rather than individual commits, this might be an easy place to get stuck. If anyone else has this problem, try rewording a commit (**not** a merge commit) that changes a file *inside one of the packages*. package.json should always work, but probably any file included in the package should also work.

**Comment by qrosmeli** [comment]
@richardkazuomiller your comment really help me. I was trying to get a major bump with a commit on a shared folder for all packages. As soon I did a silly change with a commit **inside** a package it recognized as a major change. 

Commit: `git commit -a -m feat!: Dummy breaking change`
Command: `lerna version --no-push --conventional-commits --changelog-preset conventionalcommits --no-changelog`

(We do not want the Angular preset, to be able to use the `conventionalcommits` preset you need to install the `conventional-changelog-conventionalcommits` package)

This has been the ONLY way I have been able to get a Major bump. I have been struggling with this for hours.

**Comment by richardkazuomiller** [comment]
@qrosmeli That's great! I'm glad I could help 😸

**Comment by beamery-tomht** [comment]
> If your package is in major version zero (eg. v0.x.y), the semver rules are shifted over, so breaking changes happen in v0.x.0, and features & patches happen in v0.0.y.

@privatenumber OH MY @£$@$ this has been driving me mad! Thanks so much for sharing this information!!

**Comment by andresilva-cc** [comment]
> If your package is in major version zero (eg. `v0.x.y`), the semver rules are shifted over, so _breaking changes_ happen in `v0.x.0`, and _features & patches_ happen in `v0.0.y`.
> 
> Therefore, a breaking change in a package that is in 0.x.y, will only bump _x_.
> 
> I wish this was better documented, but it's in the [npm docs](https://docs.npmjs.com/cli/v6/using-npm/semver#:~:text=%5E0.2.3%20%3A%3D%20%3E%3D0.2.3%20%3C0.3.0):
> 
> > ^1.2.3 := >=1.2.3 <2.0.0
> > ^0.2.3 := >=0.2.3 <0.3.0
> > ^0.0.3 := >=0.0.3 <0.0.4
> 
> Requested for improved documentation around this [here](https://github.com/npm/node-semver/issues/344).

And how do I get out of `0.x.y` version using the `--conventional-commits` option? Or do I need to bump it manually with `lerna version` (without using `--conventional-commits`?

**Comment by amaralc** [comment]
> @richardkazuomiller your comment really help me. I was trying to get a major bump with a commit on a shared folder for all packages. As soon I did a silly change with a commit **inside** a package it recognized as a major change.
> 
> Commit: `git commit -a -m 'feat!: Dummy breaking change'` Command: `lerna version --no-push --conventional-commits --changelog-preset conventionalcommits --no-changelog`
> 
> (We do not want the Angular preset, to be able to use the `conventionalcommits` preset you need to install the `conventional-changelog-conventionalcommits` package)
> 
> This has been the ONLY way I have been able to get a Major bump. I have been struggling with this for hours.

Thank you so much @qrosmeli and @richardkazuomiller . Passing `--changelog-preset conventionalcommits` was the only way it worked for me too.

**Comment by Allensy** [comment]
I managed to make it work!

TL;DR: 
Just add `BREAKING CHANGE: <description>}` footer at the bottom of your commit message.

I managed to resolve that using breaking change footer convention in the commit message: 

```
feat(lang)!: changed property name
BREAKING CHANGE: Changed property name `action` to `callback`
```
As you can also see here: https://www.conventionalcommits.org/en/v1.0.0/#specification
The `!` doesn't signal conventional commits that there is a breaking change, it's only used to draw attention.
The `BREAKING CHANGE` footer on the other hand actually does that. 
So running  `lerna version --conventional-commits` or `lerna publish --conventional-commits` will actually bump major versions for the affected files.

**Comment by aasen** [comment]
I read it differently, as it says `or appends a ! after the type/scope` so I would expect the **!** be enough. But I agree adding `BREAKING CHANGE: <description>}` works.

<img width="887" alt="Screenshot 2023-03-14 at 14 50 02" src="https://user-images.githubusercontent.com/484133/225022143-09ddd7d8-4bdd-4702-a354-fb6a26731e9b.png">

**Comment by ohare93** [comment]
> @richardkazuomiller your comment really help me. I was trying to get a major bump with a commit on a shared folder for all packages. As soon I did a silly change with a commit **inside** a package it recognized as a major change.
> 
> Commit: `git commit -a -m 'feat!: Dummy breaking change'` Command: `lerna version --no-push --conventional-commits --changelog-preset conventionalcommits --no-changelog`
> 
> (We do not want the Angular preset, to be able to use the `conventionalcommits` preset you need to install the `conventional-changelog-conventionalcommits` package)
> 
> This has been the ONLY way I have been able to get a Major bump. I have been struggling with this for hours.

Didn't work for me 🤔 even with a commit message of
```
feat!: lerna major bump

BREAKING CHANGE: stuff
```
Just ended up manually doing a major release to 1.0.0 via `lerna version major` 🤷‍♂️

**Comment by Klaasvaak** [comment]
To my understanding adding the following to lerna.json will respect the `!` to trigger a breaking change.

```
    "changelogPreset": {
        "name": "conventionalcommits"
    }
```

However there's also this option for lerna.json version command which does not seem to respect the `!` to trigger a breaking change:

```
    "command": {
        "version": {
            "conventionalCommits": true
        }
    },
```

Could anyone explain the difference because this is very confusing.

**Comment by maribethb** [comment]
@Klaasvaak I opened a PR to update the documentation about this issue because it also tripped up my team. My PR is https://github.com/lerna/lerna/pull/4189

It hasn't been assigned to a lerna team member yet so I'm not sure if it will be accepted, but it may help answer your question about the difference in the two flags. If it doesn't, please let me know as that was my entire goal behind the PR. :)

**Comment by JamesHenry** [comment]
Hi Folks,

I'm just doing some housekeeping and came across this old issue with tons of good discussion. With Lerna being such a well-established tool with many, many years of behaving in a particular way (by default), it felt most prudent to resolve this by capturing the behaviour very clearly in e2e tests to complement the documentation update that was made in the previously mentioned PR, so I've done that here:

https://github.com/lerna/lerna/pull/4312

I will therefore close this one as for better or worse this is how Lerna appears to have always worked by default, and there is the documented way to customize it.

Many thanks!
